packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -564,51 +564,47 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
-                    throw error;
-                }
-                this.logger.sendErrorEvent({ eventName: "TreeLatest_FallBackToGetRequest" }, error);
-                return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
-            }
-        } else {
-            return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
-        }
-    }
-
-    private async fetchSnapshotWithTimeout(snapshotOptions: ISnapshotOptions, tokenFetchOptions: TokenFetchOptions, usePost: boolean) {
-        assert(snapshotOptions.timeout !== undefined, "Timeout should be provided for setting a limit");
-        const abortController = new AbortController();
-        const timeout = setTimeout(
-            () => {
-                clearTimeout(timeout);
-                abortController.abort();
-            },
-            snapshotOptions.timeout,
-        );
         try {
             const odspSnapshot = await this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, usePost, abortController);
             return odspSnapshot;
         } catch (error) {
+            const errorType = error.errorType;
+            // If we have 401/403 error then we don't need to fallback to get call for Trees/Latest. Just throw so that we get a refreshed
+            // token for once.
+            if (errorType === DriverErrorType.authorizationError || errorType === DriverErrorType.incorrectServerResponse) {
+                throw error;
+            }
             // If the snapshot size is too big and the host specified the size limitation, then don't try to fetch the snapshot again.
-            if (error.errorType === OdspErrorType.snapshotTooBig && this.hostPolicy.snapshotOptions?.mds !== undefined) {
+            if ((error.errorType === OdspErrorType.snapshotTooBig && this.hostPolicy.snapshotOptions?.mds !== undefined) && (this.hostPolicy.summarizerClient !== true)) {
                 throw error;
             }
             // If the first snapshot request was with blobs and we either timed out or the size was too big, then try to fetch without blobs.
             if ((error.errorType === OdspErrorType.snapshotTooBig || error.errorType === OdspErrorType.fetchTimeout) && snapshotOptions.blobs) {
                 const snapshotOptionsWithoutBlobs: ISnapshotOptions = { ...snapshotOptions, blobs: 0, mds: undefined };
                 return this.fetchSnapshotCore(snapshotOptionsWithoutBlobs, tokenFetchOptions, usePost);
             }
+            // If we used the post request first time and got a 400 error from server, then try with a GET request.
+            if (usePost && error.statusCode === 400) {
+                this.logger.sendErrorEvent({ eventName: "TreeLatest_FallBackToGetRequest" }, error);
+                return this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, false);
```

Comment:
Are you sure this should not go through most of the logic of this function?
I.e. we can hit same OdspErrorType.snapshotTooBig when doing usePost = false and need to handle it the same way as this function does above, no?

Reply:
Actually it does go through all the logic of this function in case usePost is false. In try we pass value of usePost and not just true.

Reply:
What I mean is that if we started with usePost = true, we may get 404, and then we call fetchSnapshotCore(). if we get OdspErrorType.snapshotTooBig there (in fetchSnapshotCore), we fail, and never run the logic above